source/controlTypes/__init__.py

```diff
@@ -0,0 +1,236 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2007-2021 NV Access Limited, Babbage B.V.
+
+from buildVersion import version_year
+
+# imports used in __all__
+from .isCurrent import IsCurrent
+from .processing import OutputReason, processAndLabelStates, processNegativeStates, processPositiveStates
+from .role import Role, silentRolesOnFocus, silentValuesForRoles, _roleLabels
+from .state import State, STATES_SORTED, negativeStateLabels, _stateLabels
+
+
+__all__ = [
+	"IsCurrent",
+	"OutputReason",
+	"processAndLabelStates",
+	"processNegativeStates",
+	"processPositiveStates",
+	"Role",
+	"silentRolesOnFocus",
+	"silentValuesForRoles",
+	"State",
+	"STATES_SORTED",
+	"negativeStateLabels",
+]
+
+
+# Added to maintain backwards compatibility, marked for deprecation to be removed in 2022.1
+# usages to be replaced by Role.*.displayString and State.*.displayString
+if version_year < 2022:

```

Comment:
I think we should be careful to introduce automatic deprecation like this. It would be better to remove the code in 2022 anyway. Furthermore, in the current state in core, this would just break as soon as the version is updated to 2022.x.

Reply:
The code will be removed in 2022. Preparing for code removal in 2022 is the sole purpose of this PR, if we need to delay the deprecation we can simply adjust the `if` condition with a later version.

Reply:
For what its worth I do share @leonardder 's concerns here.  With this change it is extremely easy to remove parts of the code unintentionally. Normally when the new development cycle begins update of `version_year` is one of the first things done and then we can deprecate whatever needs to be deprecated piece by piece. Making backward compatibility code dependent on the value  of `version_year` forces us to one of the following:
- Deprecate everything at the beginning of the 2022.1 dev cycle in one  piece  - making it unnecessarily harder to detect regressions
- Wait with the `version_year` upgrade until all deprecation's are done - while it is not problematic as such this seems to be rather unnecessary.

If we really need to make this back compat stuff automatic then I would suggest to use `BACK_COMPAT_TO` from the `addonAPIVersion` as it is incremented just after first beta is released.

Reply:
Ah, you are right `BACK_COMPAT_TO` would definitely be the better option here. I was not thinking about when exactly we would be updating `version_year`.

Reply:
It's less about being automatic, more about being self-documenting. This code should be removed or fixed before it becomes an issue.

Reply:
I think there is a little confusion here. It is being marked as deprecated (_"deprecation is the discouragement of use of some terminology, feature, design, or practice, typically because it has been superseded or is no longer considered efficient or safe, without completely removing it or prohibiting its use")_ as of 2021.2. The goal is to remove it and prohibit it's use in the 2022.1 release. The earlier in the release we can make remove access to deprecated code, the earlier addon are aware of the change. With this mechanism, the code is effectively removed at the start of the dev cycle, as soon as we update the version number. When code is deprecated, addons should start the process of removing dependence on it. This mechanism allows addon's to test that they don't rely on the deprecations, by manually updating the version number used in NVDA.

When we get to 2022.1 we would simply remove this code.

This mechanism does mean that all core code needs to be updated to use the replacement approach. I'd say that was true anyway.

> If we really need to make this back compat stuff automatic then I would suggest to use BACK_COMPAT_TO from the addonAPIVersion as it is incremented just before first beta is released.

That is when we commit to the API, we want to have been "living with" the new API for the duration of the release dev cycle first so that any issues can be discovered. Committing to the API removals should happen as early in the release dev cycle as possible. Yes, this results in no warning to uses of addons that didn't note deprecations in prior releases. This is to be expected when running alpha software.

Reply:
As long as the entire NVDA's code base is updated to the enums *before* 2022.1 dev cycle begins and this is only to stop add-ons from accessing these old constants I'm fine with `version_year` being used here. For previous backward compatibility breaking releases we used to remove deprecate code at the same time when the code base was updated with the new approach and that prompted my comment.

Reply:
>As long as the entire NVDA's code base is updated to the enums before 2022.1 dev cycle begins

@lukaszgo1 yes, we have a mechanism for this, a new issue with a milestone has been created. If for some reason that work can't happen, we have options:
- postpone updating the version number
- change these checks to a later version number

> For previous backward compatibility breaking releases we used to remove deprecate code at the same time when the code base was updated with the new approach and that prompted my comment.

For several other enum updates we have converted the internal usages of the deprecated constants at the same time as introducing the enum. For this PR, we won't do that immediately. NVDA will continue to use the constants for a short time. This will help to ensure that this change is actually backwards compatible.

Once the backwards compatibility is confirmed, we can go ahead with removing internal usages of the newly deprecated constants. Then finally in 2022.1 we can remove the deprecated code entirely.